test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs

```diff
@@ -137,6 +137,213 @@ public EndToEndTests(ServerFixture serverFixture, ITestOutputHelper output) : ba
+                        receiveTcs.TrySetResult(Encoding.UTF8.GetString(data));
+                        return Task.CompletedTask;
+                    };
+                    connection.Closed += exception =>
+                    {
+                        logger.LogInformation("Connection closed");
+                        if (exception != null)
+                        {
+                            receiveTcs.TrySetException(exception);
+                            closeTcs.TrySetException(exception);
+                        }
+                        else
+                        {
+                            receiveTcs.TrySetResult(null);
+                            closeTcs.TrySetResult(null);
+                        }
+                    };
+
+                    logger.LogInformation("Starting connection to {url}", url);
+                    await connection.StartAsync().OrTimeout();
+                    logger.LogInformation("Started connection to {url}", url);
+
+                    var bytes = Encoding.UTF8.GetBytes(firstMessage);
+                    logger.LogInformation("Sending {length} byte message", bytes.Length);
+                    await connection.SendAsync(bytes).OrTimeout();
+                    logger.LogInformation("Sent message", bytes.Length);
+
+                    logger.LogInformation("Receiving first message");
+                    Assert.Equal(firstMessage, await receiveTcs.Task.OrTimeout());
+                    logger.LogInformation("Completed first receive");
+
+                    receiveTcs = new TaskCompletionSource<string>();
+
+                    bytes = Encoding.UTF8.GetBytes(secondMessage);
+                    logger.LogInformation("Sending {length} byte message", bytes.Length);
+                    await connection.SendAsync(bytes).OrTimeout();
+                    logger.LogInformation("Sent second message", bytes.Length);
+
+                    logger.LogInformation("Receiving second message");
+                    Assert.Equal(secondMessage, await receiveTcs.Task.OrTimeout());
+                    logger.LogInformation("Completed second receive");
+
+                    bytes = Encoding.UTF8.GetBytes(closeMessage);
+                    logger.LogInformation("Sending {length} byte message", bytes.Length);
+                    await connection.SendAsync(bytes).OrTimeout();
+                    logger.LogInformation("Sent Close message", bytes.Length);
+                    await closeTcs.Task.OrTimeout();
+                }
+                finally
+                {
+                    logger.LogInformation("Disposing Connection");
+                    await connection.DisposeAsync().OrTimeout();
+                    logger.LogInformation("Disposed Connection");
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ConnectionCanSendAndReceiveMultipleMessagesMultipleConnections()
```

Comment:
This test is necessarily long. The scenario is kind of involved. But yeah, it's a lot